File: app/cpace.js
// SPDX-License-Identifier: GPL-3.0-or-later
// CPACE-X25519-SHA512 (draft-irtf-cfrg-cpace-21, initiator-responder mode), on WebCrypto alone.
//
// The PAKE behind joining a shared pane with a meeting code and a four-digit PIN (card #97EG).
// A PIN is too small to hash into a key — anyone holding a transcript could try all ten thousand
// offline — so CPace turns it into a curve generator and runs Diffie-Hellman on that: each live
// attempt tests one PIN, and an observer tests none. The guest's browser is always the initiator.
//
// X25519 is the only group operation, and WebCrypto has it. The one thing it lacks is Elligator2
// (hashing the PIN onto the curve), which is a few lines of BigInt arithmetic mod 2^255 - 19.
// No dependencies on purpose, like app/noise.js. The desktop half is remote/cpace.py;
// tests/cpace_vectors.mjs checks this file against the draft's vectors and against the Python.
//
// ISK is only the shared secret. The caller must check key-confirmation tags derived from it
// before trusting it; that is the wire protocol's job, not this file's.

const subtle = globalThis.crypto.subtle;
const enc = new TextEncoder();

const DSI = enc.encode('CPace255');          // G_X25519.DSI
const DSI_ISK = enc.encode('CPace255_ISK');
const FIELD_BYTES = 32;
const HASH_BLOCK = 128;                      // SHA-512's input block size, H.s_in_bytes

const P = (1n << 255n) - 19n;                // Curve25519's field
const A = 486662n;                           // v^2 = u^3 + A u^2 + u
const Z = 2n;                                // Elligator2's non-square for this field

export class CPaceError extends Error {}

function concat(...parts) {
  const total = parts.reduce((n, p) => n + p.length, 0);
  const out = new Uint8Array(total);
  let at = 0;
  for (const p of parts) { out.set(p, at); at += p.length; }
  return out;
}

// ---- string encoding (draft appendix A.1, A.2, A.3.4) ---------------------------------------

function prependLen(data) {
  const len = [];
  let n = data.length;
  do {
    const byte = n & 0x7f;
    n = Math.floor(n / 128);
    len.push(n ? byte | 0x80 : byte);
  } while (n);
  return concat(new Uint8Array(len), data);
}

function lvCat(...parts) { return concat(...parts.map(prependLen)); }

function generatorString(prs, ci, sid) {
  // The zero padding fills SHA-512's first block with DSI and PRS, so the PIN's length does not
  // change how many blocks get hashed.
  const zpad = Math.max(0, HASH_BLOCK - 1 - prependLen(prs).length - prependLen(DSI).length);
  return lvCat(DSI, prs, new Uint8Array(zpad), ci, sid);
}

// The initiator's message always goes first.
function transcriptIr(ya, ada, yb, adb) { return concat(lvCat(ya, ada), lvCat(yb, adb)); }

// ---- the group (draft section 8.2) ----------------------------------------------------------

const mod = (x) => ((x % P) + P) % P;

function powMod(base, exp) {
  let result = 1n;
  base = mod(base);
  while (exp > 0n) {
    if (exp & 1n) result = (result * base) % P;
    base = (base * base) % P;
    exp >>= 1n;
  }
  return result;
}

// P is prime, so the inverse is x^(P-2).
const inv = (x) => powMod(x, P - 2n);

function fromLittleEndian(bytes) {
  let n = 0n;
  for (let i = bytes.length - 1; i >= 0; i--) n = (n << 8n) | BigInt(bytes[i]);
  return n;
}

function toLittleEndian(n) {
  const out = new Uint8Array(FIELD_BYTES);
  for (let i = 0; i < FIELD_BYTES; i++) { out[i] = Number(n & 0xffn); n >>= 8n; }
  return out;
}

// RFC 9380 map_to_curve_elligator2 for Curve25519, u-coordinate only (draft A.5).
function elligator2(r) {
  const denominator = mod(1n + Z * r * r);
  // 1 + 2r^2 = 0 would need -1/2 to be a square; it is not mod p, but RFC 9380 defines the case.
  const x1 = denominator ? mod(-A * inv(denominator)) : mod(-A);
  const gx1 = mod(x1 * x1 * x1 + A * x1 * x1 + x1);
  const square = powMod(gx1, (P - 1n) / 2n) !== P - 1n;   // Euler's criterion; 0 counts as square
  return square ? x1 : mod(-x1 - A);
}

// G_X25519.calculate_generator: the 32-byte u-coordinate g that the PIN picks.
export async function generator(prs, ci, sid) {
  const digest = new Uint8Array(await subtle.digest('SHA-512', generatorString(prs, ci, sid)));
  // RFC 7748 decodeUCoordinate: little endian with bit 255 cleared (draft A.4).
  const hashed = digest.slice(0, FIELD_BYTES);
  hashed[31] &= 0x7f;
  return toLittleEndian(elligator2(mod(fromLittleEndian(hashed))));
}

// A raw 32-byte X25519 scalar as a PKCS#8 key: WebCrypto will not import a raw private key.
const PKCS8_PREFIX = new Uint8Array([0x30, 0x2e, 0x02, 0x01, 0x00, 0x30, 0x05, 0x06, 0x03, 0x2b,
  0x65, 0x6e, 0x04, 0x22, 0x04, 0x20]);

async function importScalar(scalar) {
  if (!(scalar instanceof Uint8Array) || scalar.length !== FIELD_BYTES) {
    throw new CPaceError('a CPace scalar must be 32 bytes.');
  }
  return subtle.importKey('pkcs8', concat(PKCS8_PREFIX, scalar), { name: 'X25519' }, false, ['deriveBits']);
}

// G_X25519.scalar_mult_vfy: X25519(scalar, point), refusing the identity (all zeros).
// `scalar` is a CryptoKey from importScalar or 32 raw bytes.
export async function scalarMultVfy(scalar, point) {
  if (!(point instanceof Uint8Array) || point.length !== FIELD_BYTES) {
    throw new CPaceError('a CPace point must be 32 bytes.');
  }
  const key = scalar instanceof Uint8Array ? await importScalar(scalar) : scalar;
  let out;
  try {
    const pub = await subtle.importKey('raw', point, { name: 'X25519' }, false, []);
    out = new Uint8Array(await subtle.deriveBits({ name: 'X25519', public: pub }, key, 256));
  } catch {
    // Browsers (and Node) refuse a low-order point themselves; that is the same failure.
    throw new CPaceError("the peer's point is invalid.");
  }
  if (out.every((b) => b === 0)) throw new CPaceError("the peer's point is invalid.");
  return out;
}

// ---- the protocol (draft section 7.2) -------------------------------------------------------

// One side of one attempt. Make a fresh one per attempt: the scalar must never be reused.
export class CPace {
  // `scalar` is injectable only for test vectors; otherwise sample_scalar() is 32 random bytes,
  // which X25519 clamps.
  static async create({ prs, ci, sid, initiator, ad, scalar }) {
    const self = new CPace();
    self.sid = sid;
    self.initiator = !!initiator;
    self.ad = ad;
    self.key = await importScalar(scalar || globalThis.crypto.getRandomValues(new Uint8Array(FIELD_BYTES)));
    self.message = await scalarMultVfy(self.key, await generator(prs, ci, sid));
    return self;
  }

  // The 64-byte ISK. Throws CPaceError on a malformed or low-order peer point.
  async finish(peerY, peerAd) {
    if (!(peerY instanceof Uint8Array) || peerY.length !== FIELD_BYTES) {
      throw new CPaceError("the peer's point must be 32 bytes.");
    }
    const k = await scalarMultVfy(this.key, peerY);
    const transcript = this.initiator
      ? transcriptIr(this.message, this.ad, peerY, peerAd)
      : transcriptIr(peerY, peerAd, this.message, this.ad);
    return new Uint8Array(await subtle.digest('SHA-512', concat(lvCat(DSI_ISK, this.sid, k), transcript)));
  }
}

File: remote/cpace.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""CPACE-X25519-SHA512 — the PAKE behind "join with a meeting code and a PIN" (card #97EG).

Why a PAKE: a four-digit PIN is ten thousand guesses, which is nothing offline. Anything that lets
an eavesdropper (or the rendezvous server) check a guess against a recorded transcript — hashing
the PIN into a key, say — hands it the PIN in milliseconds. CPace turns the PIN into an elliptic
curve generator and runs Diffie-Hellman on it, so each party can test exactly one PIN per live
attempt and a passive observer can test none. The desktop burns the code after three failures.

Why CPace and this suite: it is the CFRG's recommended balanced PAKE, it needs one message each
way, and with X25519 its only group operation is X25519 itself — which WebCrypto has in the
browser and ``cryptography`` has here, the same way remote/noise.py splits its work. The one piece
neither library offers is Elligator2 (hashing to the curve), which is a few lines of field
arithmetic mod 2^255 - 19 on plain ints.

This follows draft-irtf-cfrg-cpace-21, initiator-responder mode, and is checked against the
draft's published X25519/SHA-512 vectors in tests/test_cpace.py. The browser half is
app/cpace.js; tests/cpace_vectors.mjs proves the two agree byte for byte.

    A (initiator: the guest)                  B (responder: the desktop)
    g = calculate_generator(PRS, CI, sid)     g = calculate_generator(PRS, CI, sid)
    Ya = X25519(ya, g)       -- Ya, ADa -->
                             <-- Yb, ADb --   Yb = X25519(yb, g)
    K = X25519(ya, Yb)                        K = X25519(yb, Ya)
    ISK = SHA-512(lv_cat(b"CPace255_ISK", sid, K) || lv_cat(Ya, ADa) || lv_cat(Yb, ADb))

ISK is only the shared secret. It does not prove the peer knew the PIN; the caller does that with
key-confirmation tags derived from it (the card's wire spec), and must not use ISK before they
check out.
"""
from __future__ import annotations

import os
from hashlib import sha512

from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey, X25519PublicKey

DSI = b"CPace255"                  # G_X25519.DSI
DSI_ISK = DSI + b"_ISK"
FIELD_BYTES = 32                   # G_X25519.field_size_bytes
FIELD_BITS = 255                   # G_X25519.field_size_bits
HASH_BLOCK = 128                   # SHA-512's input block size, H.s_in_bytes
IDENTITY = bytes(FIELD_BYTES)      # G_X25519.I

P = 2 ** 255 - 19                  # Curve25519's field
A = 486662                         # Curve25519: v^2 = u^3 + A u^2 + u
Z = 2                              # Elligator2's non-square for this field (RFC 9380, draft A.5)


class CPaceError(Exception):
    """An invalid peer message or a degenerate shared secret. Always fatal to the attempt."""


# ---- string encoding (draft appendix A.1, A.2, A.3.4) --------------------------------------

def _leb128(n: int) -> bytes:
    out = bytearray()
    while True:
        byte, n = n & 0x7F, n >> 7
        out.append(byte | (0x80 if n else 0))
        if not n:
            return bytes(out)


def prepend_len(data: bytes) -> bytes:
    return _leb128(len(data)) + data


def lv_cat(*parts: bytes) -> bytes:
    return b"".join(prepend_len(p) for p in parts)


def generator_string(prs: bytes, ci: bytes, sid: bytes) -> bytes:
    # The zero padding makes DSI and PRS fill the hash's whole first block, so how long the PIN
    # is does not change how many blocks get hashed.
    zpad = max(0, HASH_BLOCK - 1 - len(prepend_len(prs)) - len(prepend_len(DSI)))
    return lv_cat(DSI, prs, bytes(zpad), ci, sid)


def transcript_ir(ya: bytes, ada: bytes, yb: bytes, adb: bytes) -> bytes:
    """The initiator-responder transcript: the initiator's message always goes first."""
    return lv_cat(ya, ada) + lv_cat(yb, adb)


# ---- the group (draft section 8.2) ---------------------------------------------------------

def _elligator2(r: int) -> int:
    """RFC 9380 map_to_curve_elligator2 for Curve25519, u-coordinate only (draft A.5)."""
    denominator = (1 + Z * r * r) % P
    # 1 + 2r^2 = 0 would need -1/2 to be a square; it is not mod p, but RFC 9380 defines the case.
    x1 = (-A * pow(denominator, -1, P)) % P if denominator else (-A) % P
    gx1 = (x1 * x1 * x1 + A * x1 * x1 + x1) % P
    square = pow(gx1, (P - 1) // 2, P) != P - 1          # Euler's criterion; 0 counts as square
    return x1 if square else (-x1 - A) % P


def generator(prs: bytes, ci: bytes, sid: bytes) -> bytes:
    """G_X25519.calculate_generator: the 32-byte u-coordinate g that the PIN picks."""
    digest = sha512(generator_string(prs, ci, sid)).digest()[:FIELD_BYTES]
    # RFC 7748 decodeUCoordinate: little endian with bit 255 cleared (draft A.4).
    r = int.from_bytes(digest, "little") & ((1 << FIELD_BITS) - 1)
    return _elligator2(r % P).to_bytes(FIELD_BYTES, "little")


def scalar_mult_vfy(scalar: bytes, point: bytes) -> bytes:
    """G_X25519.scalar_mult_vfy: X25519(scalar, point), refusing the identity (all zeros)."""
    if len(scalar) != FIELD_BYTES:
        raise CPaceError("a CPace scalar must be 32 bytes.")
    if len(point) != FIELD_BYTES:
        raise CPaceError("a CPace point must be 32 bytes.")
    try:
        out = X25519PrivateKey.from_private_bytes(scalar).exchange(X25519PublicKey.from_public_bytes(point))
    except Exception as exc:  # OpenSSL refuses a low-order point itself
        raise CPaceError("the peer's point is invalid.") from exc
    if out == IDENTITY:
        raise CPaceError("the peer's point is invalid.")
    return out


# ---- the protocol (draft section 7.2) ------------------------------------------------------

class CPace:
    """One side of one attempt. Make a fresh one per attempt: the scalar must never be reused."""

    def __init__(self, prs: bytes, ci: bytes, sid: bytes, *, initiator: bool, ad: bytes,
                 scalar: bytes | None = None):
        self.sid = sid
        self.initiator = initiator
        self.ad = ad
        # sample_scalar() is 32 random bytes; X25519 clamps them. Injectable only for test vectors.
        self._scalar = os.urandom(FIELD_BYTES) if scalar is None else scalar
        if len(self._scalar) != FIELD_BYTES:
            raise CPaceError("a CPace scalar must be 32 bytes.")
        self.message = scalar_mult_vfy(self._scalar, generator(prs, ci, sid))

    def finish(self, peer_y: bytes, peer_ad: bytes) -> bytes:
        """The 64-byte ISK. Raises CPaceError on a malformed or low-order peer point."""
        if len(peer_y) != FIELD_BYTES:
            raise CPaceError("the peer's point must be 32 bytes.")
        k = scalar_mult_vfy(self._scalar, peer_y)
        if self.initiator:
            transcript = transcript_ir(self.message, self.ad, peer_y, peer_ad)
        else:
            transcript = transcript_ir(peer_y, peer_ad, self.message, self.ad)
        return sha512(lv_cat(DSI_ISK, self.sid, k) + transcript).digest()
